design/continuous-delivery.md

```diff
@@ -0,0 +1,578 @@
+
+For deployment:
+
+* **CloudFormation Role**: IAM role which allows the CloudFormation service principal to deploy stacks into the environment (this role usually has administrative privileges).
+* **Deployment Role**: IAM role which allows anyone from the deployment account to create, update and describe CloudFormation change sets and s"pass" the CloudFormation role (`iam:PassRole`).
+
+To accommodate these requirements we will make the following changes to how `cdk bootstrap` works:
+
+1. Extend the bootstrap stack to include these resources.
+2. Use explicit convention-based physical names for all resources.
+3. Allow specifying a list of *trusted accounts* from which the CloudFormation service principal can deploy to this account.
+4. Allow specifying a list of accounts from which principals can deploy to this account from the CLI.
+5. Allow specifying the managed policy to use for the deployment role (mostly it will be the administrator managed policy).
+6. Allow specifying an optional qualifier for the physical names of all resources to address bucket hijacking concerns and allow multiple bootstraps to the same environment for whatever reason.
+
+**Resource Names**
+
+We need to be able to synthesize asset locations into the templates for consumption and publishing. We also need to be able to assume a role in order to be able to publish to the environment.
+
+This means that we cannot rely on CloudFormation physical name generation since it requires accessing the account in order to resolve these names.
+
+We will employ a naming convention which encodes `account`, `region` and an optional `qualifier` (such as `cdk-account-region[-qualifier]-xxxx`). This is because not all AWS resource names are environment-local: IAM roles are account-wide and S3 buckets are global.
+
+It is important that we do not rely on hashing or parsing account and region in order to be able to support environment-agnostic stacks (in which case "account" resolves to `{ "Ref": "AWS::AccountId" }`, etc.
+
+In order to address the risk of S3 bucket hijacking, we need to be ale to support an optional `qualifier` postfix. This means that we need to allow users to specify this qualifier when they define the Stack's `env`. Perhaps we need to encode this into `aws://account/region[/qualifier]`
+
+> **Alternative considered**: one way to implement environment-specific name-spacing would have been to export the bootstrapping resources through a CloudFormation Export and then reference them using Fn::ImportValue. This would have worked for templates, but means that we would need a way to resolve import values during publishing as well (and as a result also Fn::Join, etc).
+
+**User Interface**
+
+This is the proposed API for `cdk-bootstrap`:
+
+```
+$ cdk bootstrap --profile XXX [--yes] aws://account/region --trust-account ACCOUNT_ID
+WARNING: any principal from <ACCOUNT_ID> will have administrative access to the account <account>.
+Please confirm (Y/N):
```

Comment:
What about the prompting experience that allows you to choose from the ManagedPolicies to assign to your CFN role?

Reply:
Can probably start without it and add later.